fem/bilinearform_ext.cpp

```diff
@@ -287,6 +289,417 @@ void PABilinearFormExtension::MultTranspose(const Vector &x, Vector &y) const
    }
 }
 
+// Data and methods for element-assembled bilinear forms
+EABilinearFormExtension::EABilinearFormExtension(BilinearForm *form)
+   : BilinearFormExtension(form),
+     trialFes(a->FESpace()),
+     testFes(a->FESpace())
+{
+   elem_restrict = NULL;
+   int_face_restrict_lex = NULL;
+   bdr_face_restrict_lex = NULL;
+}
+
+void EABilinearFormExtension::SetupRestrictionOperators()
+{
+   ElementDofOrdering ordering = UsesTensorBasis(*a->FESpace())?
+                                 ElementDofOrdering::LEXICOGRAPHIC:
+                                 ElementDofOrdering::NATIVE;
+   elem_restrict = trialFes->GetElementRestriction(ordering);
+   if (elem_restrict)
+   {
+      localX.SetSize(elem_restrict->Height(), Device::GetMemoryType());
+      localY.SetSize(elem_restrict->Height(), Device::GetMemoryType());
+      localY.UseDevice(true); // ensure 'localY = 0.0' is done on device
+   }
+
+   // Construct face restriction operators only if the bilinear form has
+   // interior or boundary face integrators
+   if (int_face_restrict_lex == NULL && a->GetFBFI()->Size() > 0)
+   {
+      int_face_restrict_lex = trialFes->GetFaceRestriction(
+                                 ElementDofOrdering::LEXICOGRAPHIC,
+                                 FaceType::Interior);
+      faceIntX.SetSize(int_face_restrict_lex->Height(), Device::GetMemoryType());
+      faceIntY.SetSize(int_face_restrict_lex->Height(), Device::GetMemoryType());
+      faceIntY.UseDevice(true); // ensure 'faceIntY = 0.0' is done on device
+   }
+
+   if (bdr_face_restrict_lex == NULL && a->GetBFBFI()->Size() > 0)
+   {
+      bdr_face_restrict_lex = trialFes->GetFaceRestriction(
+                                 ElementDofOrdering::LEXICOGRAPHIC,
+                                 FaceType::Boundary,
+                                 L2FaceValues::SingleValued);
```

Comment:
Is this the only difference with `PABilinearFormExtension`? If so, we might want to factorize some code: we should talk about it.

Reply:
I considered it at some point, I wasn't very successful, I'll retry.

Reply:
I used `public` inheritance, to avoid these code duplications, tell me if you like it.